pytorch_lightning/experimental/plugins/secure/pysyft.py

```diff
@@ -0,0 +1,121 @@
+    def optimizer_state(optimizer: Optimizer) -> Dict[str, torch.Tensor]:
+        """
+        Returns state of an optimizer. Allows for syncing/collating optimizer state from processes in custom
+        plugins.
+        """
+        return optimizer.state_dict().get(request_block=True, delete_obj=False)
+
+    @staticmethod
+    def save_function(trainer, filepath: str, save_weights_only: bool) -> None:
+        model_ref = trainer.lightning_module
+        sy_model = model_ref.get_model()
+        sy_model.device = model_ref.device
+        sy_model.hparams = model_ref.hparams
+        sy_model.on_save_checkpoint = model_ref.on_save_checkpoint
+        trainer.training_type_plugin.model = sy_model
+        trainer.accelerator.optimizer_state = PySyftPlugin.optimizer_state
+        trainer.save_checkpoint(filepath, save_weights_only)
+        trainer.training_type_plugin.model = model_ref
+
+
+if _PYSYFT_AVAILABLE:
+
+    from types import ModuleType
+    from typing import Any, Union
+
+    import syft as sy
+
+    from pytorch_lightning.core.lightning import LightningModule
+
+    SyModuleProxyType = Union[ModuleType, Module]
+    SyModelProxyType = Union[Module, sy.Module]
+
+    # cant use lib_ast during test search time
+    TorchTensorPointerType = Any  # sy.lib_ast.torch.Tensor.pointer_type
+    SyTensorProxyType = Union[torch.Tensor, TorchTensorPointerType]  # type: ignore
+
+    class SyLightningModule(LightningModule):
+
+        def __init__(self, download_back: bool = False, run_locally: bool = False) -> None:
+            super().__init__()
+            # Those are helpers to easily work with `sy.Module`
+            self.duet = sy.client_cache["duet"]
+            self.remote_torch = sy.client_cache["duet"].torch
+            self.local_torch = globals()["torch"]
+            self.download_back = download_back
+            self.run_locally = run_locally
+
+        def setup(self, stage_name: str):
+            self.get = self.module.get
+            self.send = self.module.send
+            self.send_model()
+
+        def is_remote(self) -> bool:
+            # Training / Evaluation is done remotely and Testing is done locally unless run_locally is True
+            if self.run_locally or (not self.trainer.training and self.trainer.evaluation_loop.testing):
+                return False
+            return True
+
+        @property
+        def torch(self) -> SyModuleProxyType:
+            return self.remote_torch if self.is_remote() else self.local_torch
+
+        @property
+        def model(self) -> SyModelProxyType:
+            if self.is_remote():
+                return self.remote_model
+            else:
+                if self.download_back:
+                    return self.get_model()
+                else:
+                    return self.module
+
+        def send_model(self) -> None:
+            self.remote_model = self.module.send(self.duet)
+
+        def get_model(self) -> type(Module):  # type: ignore
```

Comment:
why is the return type not just `Module`?

Reply:
No idea. Not my code ahah. Remove it.